### PR TITLE
Add ReplaySafeLoggerFactory for context wrappers

### LINKED ISSUE #497: Make `TaskOrchestrationContext.LoggerFactory` public
```cs
// TaskOrchestrationContext.cs
protected abstract ILoggerFactory LoggerFactory { get; }
```

This is a problem in scenarios where the context is wrapped by another context implementation (concrete use case for me is implementing a cancellable orchestrator which wraps the original context). My current workaround is to override the methods that actually use this property (namely the `CreateReplaySafeLogger` overloads), but that's a dependency on implementation details. The proper solution in this case would be to just return the `LoggerFactory` of the wrapped context, if it was made public. `LoggerFactory` is the only member of `TaskOrchestrationContext` that is not public.

### PR DESCRIPTION
**Pull request by torosent**
# Summary
## What changed?
- Add a new public virtual `TaskOrchestrationContext.ReplaySafeLoggerFactory` property.
- Add a small internal `ReplaySafeLoggerFactoryImpl` wrapper that implements `ILoggerFactory` and delegates `CreateLogger(string)` to the existing replay-safe logger path.
- Keep the existing `protected abstract LoggerFactory` contract unchanged.
- Add a changelog entry for the new API.

## Why is this change needed?
- Issue #497 highlights a wrapper scenario where a derived `TaskOrchestrationContext` needs to delegate `LoggerFactory` to an inner context, but C# protected access rules do not allow `_inner.LoggerFactory`.
- Simply making `LoggerFactory` public would weaken the current pit-of-success design by letting customers create raw loggers that bypass replay-safe logging and produce duplicate logs during orchestration replay.
- This design solves the wrapper scenario without exposing the raw factory: wrappers can safely do `protected override ILoggerFactory LoggerFactory => inner.ReplaySafeLoggerFactory;`.
- Architecturally, the new property stays on the abstraction layer and reuses the existing `CreateReplaySafeLogger(...)` implementation, so the replay-safety behavior remains centralized instead of duplicating logging logic in wrappers or derived types.

## Issues / work items
- Resolves #497
- Related #

---

# Project checklist
- [x] Release notes are not required for the next release
  - [ ] Otherwise: Notes added to `release_notes.md`
- [x] Backport is not required
  - [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
- [ ] All required tests have been added/updated (unit tests, E2E tests)
- [ ] Breaking change?
  - [ ] If yes:
    - Impact:
    - Migration guidance:
---

# AI-assisted code disclosure (required)
## Was an AI tool used? (select one)
- [ ] No
- [x] Yes, AI helped write parts of this PR (e.g., GitHub Copilot)
- [ ] Yes, an AI agent generated most of this PR

If AI was used:
- Tool(s): GitHub Copilot CLI / GPT-5.4
- AI-assisted areas/files: `src/Abstractions/TaskOrchestrationContext.cs`, `CHANGELOG.md`
- What you changed after AI output: Rejected the initial direct visibility change that made `LoggerFactory` public, then replaced it with the safer `ReplaySafeLoggerFactory` design after replay-safety review.

AI verification (required if AI was used):
- [x] I understand the code and can explain it
- [x] I verified referenced APIs/types exist and are correct
- [x] I reviewed edge cases/failure paths (timeouts, retries, cancellation, exceptions)
- [x] I reviewed concurrency/async behavior
- [x] I checked for unintended breaking or behavior changes

---

# Testing
## Automated tests
- Result: Passed (`dotnet build Microsoft.DurableTask.sln` and `dotnet test Microsoft.DurableTask.sln --no-build`)

## Manual validation (only if runtime/behavior changed)
- Environment (OS, .NET version, components): macOS (Darwin), repository SDK/tooling from `global.json`
- Steps + observed results:
  1. Built `Microsoft.DurableTask.sln` successfully with 0 errors.
  2. Ran the full solution test suite successfully.
  3. Reviewed the wrapper architecture to confirm the public API exposes only replay-safe logger creation while the raw `LoggerFactory` remains protected.
- Evidence (optional):

---

# Notes for reviewers
- The key design point is that raw `LoggerFactory` intentionally remains `protected`; the new public API exposes only a replay-safe wrapper so wrapper contexts can opt into the existing replay-safe logging path.

### CODE REVIEW

File: src/Abstractions/TaskOrchestrationContext.cs [hunk 1]
```diff
@@ -76,6 +78,25 @@ public abstract class TaskOrchestrationContext
...
+    /// duplicate log messages during replay. This is equivalent to calling
+    /// <see cref="CreateReplaySafeLogger(string)"/> for each logger category.
+    /// </para>
+    /// <para>
+    /// Context wrapper implementations can delegate <see cref="LoggerFactory"/> to this property
+    /// on the inner context: <c>protected override ILoggerFactory LoggerFactory =&gt; inner.ReplaySafeLoggerFactory;</c>.
+    /// </para>
+    /// </remarks>
+    public virtual ILoggerFactory ReplaySafeLoggerFactory
+        => this.replaySafeLoggerFactory ??= new ReplaySafeLoggerFactoryImpl(this);
```
Comment (YunchuWang) [inline src/Abstractions/TaskOrchestrationContext.cs:98]: **[F1] Double-wrapped replay-safe loggers  Medium / P2**

When a wrapper does `protected override ILoggerFactory LoggerFactory => inner.ReplaySafeLoggerFactory;`, the base class `CreateReplaySafeLogger(string)` calls `this.LoggerFactory.CreateLogger(name)` which returns a `ReplaySafeLogger` from the inner context, then wraps it **again** in another `ReplaySafeLogger`. Every `Log()` call checks `IsReplaying` twice.

**Call chain for wrapper usage:**
1. `wrapper.CreateReplaySafeLogger("Foo")`
2.  `new ReplaySafeLogger(wrapper, wrapper.LoggerFactory.CreateLogger("Foo"))`
3.  `wrapper.LoggerFactory` returns `inner.ReplaySafeLoggerFactory`
4.  `ReplaySafeLoggerFactoryImpl.CreateLogger("Foo")`  `inner.CreateReplaySafeLogger("Foo")`
5.  `new ReplaySafeLogger(inner, inner.LoggerFactory.CreateLogger("Foo"))`  first wrap
6.  Back at step 2: `new ReplaySafeLogger(wrapper, <already-wrapped>)`  second wrap

This is functionally correct (replay suppression is idempotent) but adds unnecessary indirection.

**Recommendation**: `ReplaySafeLoggerFactoryImpl.CreateLogger` could call `this.context.LoggerFactory.CreateLogger(categoryName)` directly (nested classes can access protected members of their enclosing class) instead of going through `CreateReplaySafeLogger`, since the factory wrapper itself already provides the replay-safe guarantee. Alternatively, document this double-wrap as intentional.

**Review by torosent — COMMENTED** [review]
All findings addressed in commit d6c8df4:

**F1 (Double-wrapped replay-safe loggers)**: Added `GetUnwrappedLoggerFactory()` which peels through any `ReplaySafeLoggerFactoryImpl` layers before creating loggers. Also added a max-depth guard (10 iterations) to protect against infinite loops if a wrapper mistakenly points `LoggerFactory` at `this.ReplaySafeLoggerFactory` instead of the inner context. Added regression tests for both double-wrap prevention and cycle detection.

**F3 (Sample wrapper comment)**: Added a comment in the sample wrapper noting that virtual methods with default base implementations delegate through abstract overrides and don't need explicit forwarding. Also removed the unnecessary `Entities` override.

**F4 (Moq inconsistency)**: Replaced `Mock<ILoggerProvider>` with the hand-rolled `TrackingLoggerProvider`, removing the Moq dependency from this test class.

**F5 (Unused using)**: Removed `using Microsoft.DurableTask.Entities`.

**F6 (net6.0 targeting)**: No change needed — follows existing convention.

File: src/Abstractions/TaskOrchestrationContext.cs (hunk 1 above)
Comment (torosent) [reply src/Abstractions/TaskOrchestrationContext.cs:98]: Addressed in d6c8df4. Added `GetUnwrappedLoggerFactory()` which peels through `ReplaySafeLoggerFactoryImpl` layers to reach the real `ILoggerFactory` before creating loggers. This avoids the double-wrap while keeping the factory's `CreateLogger` clean. Also added a max-depth guard (10 iterations) to catch infinite loops from misconfigured wrappers, plus regression tests for both scenarios.

File: test/Abstractions.Tests/TaskOrchestrationContextReplaySafeLoggerFactoryTests.cs [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+        public TrackingLoggerFactory(TrackingLoggerProvider provider)
+        {
+            this.provider = provider;
+        }
+
+        public int AddProviderCallCount { get; private set; }
+
+        public int DisposeCallCount { get; private set; }
+
+        public void AddProvider(ILoggerProvider provider)
```
[the remark this answers is not in the document]
Comment (torosent) [reply test/Abstractions.Tests/TaskOrchestrationContextReplaySafeLoggerFactoryTests.cs]: Fixed in f888b35 — renamed the parameter from `provider` to `loggerProvider` to avoid shadowing the field.

File: src/Abstractions/TaskOrchestrationContext.cs [hunk 3]
```diff
@@ -445,7 +466,31 @@ public virtual int CompareVersionTo(string version)
...
-    class ReplaySafeLogger : ILogger
+    ILoggerFactory GetUnwrappedLoggerFactory()
+    {
+        ILoggerFactory loggerFactory = this.LoggerFactory;
+        int depth = 0;
+
+        // When a wrapper context delegates LoggerFactory to inner.ReplaySafeLoggerFactory,
+        // the returned factory is already a ReplaySafeLoggerFactoryImpl. Unwrap it to avoid
+        // double-wrapping loggers with redundant replay-safe checks.
+        while (loggerFactory is ReplaySafeLoggerFactoryImpl replaySafeLoggerFactory)
```
[the remark this answers is not in the document]
Comment (torosent) [reply src/Abstractions/TaskOrchestrationContext.cs]: Fixed in b118da9 — renamed the pattern variable to `wrappedFactory` to avoid shadowing the `replaySafeLoggerFactory` field.

**Comment by torosent** [comment]
@copilot open a new pull request to apply changes based on the comments in [this thread](https://github.com/microsoft/durabletask-dotnet/pull/670#pullrequestreview-3949553549)

File: src/Abstractions/TaskOrchestrationContext.cs [hunk 4]
```diff
@@ -445,7 +466,31 @@ public virtual int CompareVersionTo(string version)
...
+        // double-wrapping loggers with redundant replay-safe checks.
+        while (loggerFactory is ReplaySafeLoggerFactoryImpl wrappedFactory)
+        {
+            if (++depth > 10)
+            {
+                throw new InvalidOperationException(
+                    "Cycle detected while unwrapping ReplaySafeLoggerFactory. " +
+                    "Ensure the wrapper's LoggerFactory property delegates to the inner context's " +
+                    "ReplaySafeLoggerFactory (e.g., 'inner.ReplaySafeLoggerFactory'), not 'this.ReplaySafeLoggerFactory'.");
+            }
```
[the remark this answers is not in the document]
Comment (torosent) [reply src/Abstractions/TaskOrchestrationContext.cs:485]: Not fixing

File: Microsoft.DurableTask.sln [hunk 5]
```diff
@@ -111,18 +111,41 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "ExportHistory.Tests", "test
...
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "ReplaySafeLoggerFactorySample", "samples\ReplaySafeLoggerFactorySample\ReplaySafeLoggerFactorySample.csproj", "{8E7BECBC-7226-4778-B8F2-8EBDFF0D3BA4}"
+EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Grpc", "Grpc", "{51A52603-541D-DE3F-2825-C80F9EE6C532}"
+EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "AzureManaged", "AzureManaged", "{D4587EC0-1B16-8420-7502-A967139249D4}"
+EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Grpc", "Grpc", "{3B8F957E-7773-4C0C-ACD7-91A1591D9312}"
+EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "AzureManaged", "AzureManaged", "{53193780-CD18-2643-6953-C26F59EAEDF5}"
+EndProject
```
[the remark this answers is not in the document]
Comment (torosent) [reply Microsoft.DurableTask.sln]: fixed

File: Microsoft.DurableTask.sln [hunk 6]
```diff
@@ -111,18 +111,41 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "ExportHistory.Tests", "test
...
 Global
 	GlobalSection(SolutionConfigurationPlatforms) = preSolution
 		Debug|Any CPU = Debug|Any CPU
+		Debug|x64 = Debug|x64
+		Debug|x86 = Debug|x86
 		Release|Any CPU = Release|Any CPU
+		Release|x64 = Release|x64
+		Release|x86 = Release|x86
```
[the remark this answers is not in the document]
Comment (torosent) [reply Microsoft.DurableTask.sln]: fixed

**Review by YunchuWang — APPROVED** [review]
LGTM  clean design with good test coverage. A few minor items:

**[F1] `SelfReferencingContext.loggerFactory` field is unused (Medium/P2)**
`SelfReferencingContext` stores the constructor `ILoggerFactory` parameter in a field, but `LoggerFactory` returns `this.ReplaySafeLoggerFactory` instead. The field is dead code (analyzer flags it too). Consider removing it.

**[F2] `TrackingLoggerFactory.AddProvider` parameter name (Nit)**
Rename `loggerProvider` to `provider` to match the `ILoggerFactory.AddProvider` interface declaration (analyzer warning).

**[F3] `GetInput<T>()` override signature (Nit)**
Base declares `T? GetInput<T>()` but test stubs override as `T GetInput<T>() => default!` (narrowing return type). Compare with `TaskOrchestrationContextVersionTests.cs` which correctly uses `T? GetInput<T>() where T : default`.

**Review by YunchuWang — APPROVED** [review]
